nifi-extension-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/kinesis/stream/consumev2/ConsumeKinesisStreamV2.java

```diff
@@ -0,0 +1,572 @@
+        final Optional<ShardBufferLease> maybeLease = recordBuffer.acquireBufferLease();
+
+        maybeLease.ifPresentOrElse(
+                lease -> processRecordsFromBuffer(session, lease),
+                context::yield
+        );
+    }
+
+    private void processRecordsFromBuffer(final ProcessSession session, final ShardBufferLease lease) {
+        try {
+            final List<KinesisClientRecord> records = recordBuffer.consumeRecords(lease);
+
+            if (records.isEmpty()) {
+                recordBuffer.returnBufferLease(lease);
+                return;
+            }
+
+            final String shardId = lease.shardId();
+
+            processRecordsAsRaw(session, shardId, records);
+
+            // TODO support Record Reader and Writer.
+//            if (recordProcessor == null) {
+//                processRecordsAsRaw(session, shardId, records);
+//            } else {
+//                processRecordsWithRecordProcessor(recordProcessor, session, shardId, records);
+//            }
+
+            session.adjustCounter("Records Processed", records.size(), false);
+
+            session.commitAsync(
+                    () -> {
+                        recordBuffer.commitConsumedRecords(lease);
+                        recordBuffer.returnBufferLease(lease);
+                    },
+                    __ -> {
+                        recordBuffer.rollbackConsumedRecords(lease);
+                        recordBuffer.returnBufferLease(lease);
+                    }
+            );
+        } catch (final RuntimeException e) {
+            getLogger().error("Failed to process records from Kinesis stream: {}", e.getMessage());
+            recordBuffer.rollbackConsumedRecords(lease);
+            recordBuffer.returnBufferLease(lease);
+            throw e;
+        }
+    }
+
+    private static void processRecordsAsRaw(final ProcessSession session, final String shardId, final List<KinesisClientRecord> records) {
+        for (final KinesisClientRecord record : records) {
+            final FlowFile flowFile = session.create();
+            session.putAllAttributes(flowFile, ConsumeKinesisStreamV2Attributes.fromKinesisRecord(shardId, record));
+
+            session.write(flowFile, out -> Channels.newChannel(out).write(record.data()));
+
+            session.transfer(flowFile, REL_SUCCESS);
+        }
+    }
+
+    /**
+     * An adapter between Kinesis Consumer Library and {@link RecordBuffer}.
+     */
+    private static class ConsumeKinesisRecordProcessor implements ShardRecordProcessor {
+
+        private final RecordBuffer recordBuffer;
+        private volatile ShardBufferId bufferId;
+
+        ConsumeKinesisRecordProcessor(final RecordBuffer recordBuffer) {
+            this.recordBuffer = recordBuffer;
+        }
+
+        @Override
+        public void initialize(final InitializationInput initializationInput) {
+            bufferId = recordBuffer.createBuffer(initializationInput.shardId());
```

Comment:
Suggestion: check at the construction time if the bufferId is present
```suggestion
            bufferId = Objects.requireNonNull(recordBuffer.createBuffer(initializationInput.shardId()));
```

Reply:
What I meant: it is more of a nit — this class is not usable at all if this parameter is `null` — null checks are in every function of this class, so we could have only one check in constructor instead one in every function.

The potential downside is that calls of these functions come from code that we control, whereas constructor is called indirectly by the framework.